File: cs/days/day23.cs
namespace Shunty.AoC.Days;

// https://adventofcode.com/2024/day/23 - LAN Party

public class Day23 : AocDaySolver
{
    public int DayNumber => 23;
    public string Title => "LAN Party";

    public async Task Solve()
    {
        var input = (await File.ReadAllLinesAsync(AocUtils.FindInputFile(DayNumber)))
        //var input = (await Task.FromResult(TestInput.Trim())).Split('\n').Select(s => s.Trim())
            .Select(ln => ln.Split('-'))
            .Select(s => (s[0], s[1]))
            .ToList();

        Dictionary<string, HashSet<string>> lan = [];
        foreach (var (n1, n2) in input)
        {
            if (lan.TryGetValue(n1, out var nn1))
                nn1.Add(n2);
            else
                lan.Add(n1, [n2]);

            if (lan.TryGetValue(n2, out var nn2))
                nn2.Add(n1);
            else
                lan.Add(n2, [n1]);
        }


        HashSet<(string,string,string)> tgroups = [];
        Dictionary<string, int> lanparty = [];
        foreach (var (k1,v) in lan)
        {
            foreach (var k2 in v)
            {
                var common = v.Intersect(lan[k2]).ToList();
                if (common.Count > 0)
                {
                    // Part 2
                    if (common.Count > 1)
                    {
                        List<string> lp = [k1, k2, .. common];
                        lp.Sort();
                        var lpkey = string.Join(",", lp);
                        if (lanparty.TryGetValue(lpkey, out var lpc))
                            lanparty[lpkey] = lpc+1;
                        else
                            lanparty[lpkey] = 1;
                    }

                    // Part 1
                    foreach (var k3 in common)
                    {
                        List<string> sorted = [k1, k2, k3];
                        sorted.Sort();
                        if (sorted.Any(x => x[0] == 't'))
                            tgroups.Add((sorted[0], sorted[1], sorted[2]));

                    }
                }
            }
        }

        this.ShowDayResult(1, tgroups.Count);
        var p2max = lanparty.Max(x => x.Value);
        this.ShowDayResult(2, lanparty.First(kvp => kvp.Value == p2max).Key);
    }

    private const string TestInput = """
    kh-tc
    qp-kh
    de-cg
    ka-co
    yn-aq
    qp-ub
    cg-tb
    vc-aq
    tb-ka
    wh-tc
    yn-cg
    kh-ub
    ta-co
    de-co
    tc-td
    tb-wq
    wh-td
    ta-ka
    td-qp
    aq-cg
    wq-ub
    ub-vc
    de-ta
    wq-aq
    wq-vc
    wh-yn
    ka-de
    kh-ta
    co-tc
    wh-qp
    tb-vc
    td-yn
    """;
}

File: go/days/day23.go
package days

// https://adventofcode.com/2024/day/23 - LAN Party

import (
	aoc "aoc2024/aocutils"
	"slices"
	"strings"
)

type day23 struct {
	day int
}

func Day23() {
	d := &day23{23}
	d.Run()
}

func (d *day23) Run() {
	input, err := aoc.GetDayInputLines(d.day)
	aoc.CheckError(err)

	lan := make(map[string][]string)
	for _, ln := range input {
		sp := strings.Split(strings.TrimSpace(ln), "-")
		lan[sp[0]] = append(lan[sp[0]], sp[1])
		lan[sp[1]] = append(lan[sp[1]], sp[0])
	}
	for _, v := range lan {
		slices.Sort(v)
	}

	p1 := aoc.NewSet[[3]string]()
	p2 := aoc.NewCounter[string]()
	for k1, v := range lan {
		for _, k2 := range v {
			common := aoc.Intersect(v, lan[k2])
			if len(common) > 0 {

				// Part 1
				for _, k3 := range common {
					if strings.HasPrefix(k1, "t") || strings.HasPrefix(k2, "t") || strings.HasPrefix(k3, "t") {
						connected := []string{k1, k2, k3}
						slices.Sort(connected)
						p1.Add([3]string(connected))
					}
				}

				// Part 2
				if len(common) > 1 {
					party := []string{k1, k2}
					party = append(party, common...)
					slices.Sort(party)
					partykey := strings.Join(party, ",")
					p2.Add(partykey)
				}
			}
		}
	}

	aoc.DayHeader(d.day)
	aoc.PrintResult(1, p1.Count())
	aoc.PrintResult(2, p2.MaxKey())
}
